plugin/trino-exchange/src/main/java/io/trino/plugin/exchange/FileSystemExchangeSource.java

```diff
@@ -15,116 +15,115 @@
+        for (int i = 0; i < numReaders; ++i) {
+            readers.add(exchangeStorage.createExchangeStorageReader(sourceFileQueue, maxPageStorageSize));
+        }
+        this.readers = readers.build();
     }
 
     @Override
     public CompletableFuture<Void> isBlocked()
     {
-        return NOT_BLOCKED;
+        for (ExchangeStorageReader reader : readers) {
+            if (reader.isBlocked().isDone()) {
+                return NOT_BLOCKED;
+            }
+        }
+        return toCompletableFuture(
+                nonCancellationPropagating(
+                        whenAnyComplete(readers.stream()
+                                .map(ExchangeStorageReader::isBlocked)
+                                .collect(toImmutableList()))));
     }
 
     @Override
-    public synchronized boolean isFinished()
+    public boolean isFinished()
     {
-        return closed || (!files.hasNext() && sliceInput == null);
+        if (closed) {
+            return true;
+        }
+
+        for (ExchangeStorageReader reader : readers) {
+            if (!reader.isFinished()) {
+                return false;
+            }
+        }
+        return true;
     }
 
     @Nullable
     @Override
-    public synchronized Slice read()
+    public Slice read()
     {
-        if (isFinished()) {
+        if (closed) {
             return null;
         }
 
-        if (sliceInput != null && !sliceInput.isReadable()) {
-            sliceInput.close();
-            sliceInput = null;
-        }
-
-        if (sliceInput == null) {
-            if (files.hasNext()) {
-                // TODO: implement parallel read
-                URI file = files.next();
-                Optional<SecretKey> secretKey = secretKeys.next();
+        for (ExchangeStorageReader reader : readers) {
+            if (reader.isBlocked().isDone() && !reader.isFinished()) {
                 try {
-                    sliceInput = exchangeStorage.getSliceInput(file, secretKey);
+                    return reader.read();
                 }
                 catch (IOException e) {
                     throw new UncheckedIOException(e);
                 }
             }
         }
 
-        if (sliceInput == null) {
-            return null;
-        }
-
-        if (!sliceInput.isReadable()) {
-            sliceInput.close();
-            sliceInput = null;
-            return null;
-        }
-
-        int size = sliceInput.readInt();
-        return sliceInput.readSlice(size);
+        return null;
     }
 
     @Override
-    public synchronized long getMemoryUsage()
+    public long getMemoryUsage()
     {
-        return sliceInput != null ? sliceInput.getRetainedSize() : 0;
+        long memoryUsage = 0;
+        for (ExchangeStorageReader reader : readers) {
+            memoryUsage += reader.getRetainedSize();
+        }
+        return memoryUsage;
     }
 
     @Override
-    public synchronized void close()
+    public void close()
```

Comment:
nit: While the current implementation is still correct i wonder if we should keep it as `synchronized` to avoid closing more than once?